engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementActivationBehavior.java

```diff
@@ -0,0 +1,245 @@
+import io.camunda.zeebe.engine.processing.streamprocessor.sideeffect.SideEffectQueue;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.StateWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.TypedCommandWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.Writers;
+import io.camunda.zeebe.engine.state.KeyGenerator;
+import io.camunda.zeebe.engine.state.immutable.ElementInstanceState;
+import io.camunda.zeebe.engine.state.instance.ElementInstance;
+import io.camunda.zeebe.protocol.impl.record.value.processinstance.ProcessInstanceRecord;
+import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
+import io.camunda.zeebe.protocol.record.value.BpmnElementType;
+import io.camunda.zeebe.util.Either;
+import io.camunda.zeebe.util.buffer.BufferUtil;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+public final class ElementActivationBehavior {
+
+  private final SideEffectQueue sideEffectQueue = new SideEffectQueue();
+
+  private final KeyGenerator keyGenerator;
+  private final TypedCommandWriter commandWriter;
+  private final StateWriter stateWriter;
+  private final CatchEventBehavior catchEventBehavior;
+
+  private final ElementInstanceState elementInstanceState;
+
+  public ElementActivationBehavior(
+      final KeyGenerator keyGenerator,
+      final Writers writers,
+      final CatchEventBehavior catchEventBehavior,
+      final ElementInstanceState elementInstanceState) {
+    this.keyGenerator = keyGenerator;
+    this.catchEventBehavior = catchEventBehavior;
+    this.elementInstanceState = elementInstanceState;
+    commandWriter = writers.command();
+    stateWriter = writers.state();
+  }
+
+  /**
+   * Activates the flow scopes of a given element. This is used when modifying a process instance or
+   * starting a process instance at a different place than the start event.
+   *
+   * @param processInstanceRecord the record of the process instance
+   * @param elementToActivate The element to activate
+   * @return The key of the activated element instance
+   */
+  public long activateElement(
+      final ProcessInstanceRecord processInstanceRecord,
+      final AbstractFlowElement elementToActivate) {
+
+    final var flowScopes = collectFlowScopesOfElement(elementToActivate);
+
+    final var flowScopeKey =
+        activateFlowScopes(
+            processInstanceRecord, processInstanceRecord.getProcessInstanceKey(), flowScopes);
+
+    final long elementInstanceKey =
+        activateElementByCommand(processInstanceRecord, elementToActivate, flowScopeKey);
+
+    // applying the side effects is part of creating the event subscriptions
+    sideEffectQueue.flush();
+
+    return elementInstanceKey;
+  }
+
+  private List<ExecutableFlowElement> collectFlowScopesOfElement(
+      final ExecutableFlowElement element) {
+    final List<ExecutableFlowElement> flowScopes = new ArrayList<>();
```

Comment:
🔧 I think it'd be easier to use a stack here.

Reply:
Good idea. I will use a [Deque](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Deque.html) instead. `Deque` should be preferred over `Stack` (see the note on the [JavaDoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Stack.html)).